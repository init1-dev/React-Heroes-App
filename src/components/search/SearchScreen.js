import { useNavigate, useLocation } from "react-router-dom";
import { useMemo } from "react";
import queryString from "query-string";

import { useForm } from "../../hooks/useForm";
import { getHeroesByName } from "../../selectors/getHeroesByName";
import { HeroCard } from "../hero/HeroCard";

export const SearchScreen = () => {

    const navigate = useNavigate();
    const location = useLocation();

    const { q = '' } = queryString.parse(location.search);

    const [ formValues, handleInputChange ] = useForm({
        searchText: q
    });

    const { searchText } = formValues;
    
    const heroesFiltered = useMemo( () => getHeroesByName(q), [q] );

    const handleSearch = (e) => {
        e.preventDefault();

        navigate(`?q=${ searchText }`);
    };

    return (
        <>
            <h1 className="mt-3">Search</h1>
            <hr />

            <div className="row">

                <div className="col-5">
                    {/* <h4>Buscar</h4>
                    <hr /> */}

                    <form onSubmit={ handleSearch }>
                        <input 
                            type="text" 
                            className="form-control"
                            name="searchText"
                            placeholder="Buscar un Héroe"
                            autoComplete="off"
                            value={ searchText }
                            onChange={ handleInputChange }
                        />

                        <button 
                            className="btn btn-outline-primary mt-1"
                            type="submit"
                        >
                            Buscar
                        </button>

                    </form>
                </div>

                <div className="col-7 mb-3">

                    {
                        (q.length === 0 )
                            ? <div className="alert alert-info"> Introduce un héroe a buscar </div>
                            : ( heroesFiltered.length === 0 )
                                && <div className="alert alert-danger"> No hay resultados para: { q } </div>
                    }

                    {
                        ( heroesFiltered.length > 0 )
                            && (<>
                                    <h4>Resultados</h4>
                                    <hr /> 
                                </>)
                    }

                    {
                        heroesFiltered.map( hero => (
                            <HeroCard 
                                key={ hero.id }
                                {...hero}
                            />
                        ))
                    }
                </div>

            </div>
        </>
    )
}
