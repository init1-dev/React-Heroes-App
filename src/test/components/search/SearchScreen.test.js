import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import { SearchScreen } from "../../../components/search/SearchScreen";

// Es imprescindible poner mock delante del nombre que queramos darle a la constante para evitar errores de scope
const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate
}))

describe('pruebas en <SearchScreen />', () => {
    
    test('debe mostrarse correctamente con valores por defecto', () => {
       
        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search'] }>
                <SearchScreen />
            </MemoryRouter>
        );
        
        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.alert-info').text().trim() ).toBe('Introduce un hÃ©roe a buscar');
    });
    
    test('debe mostrar a Batman y el input con el valor del queryString', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search?q=batman'] }>
                <SearchScreen />
            </MemoryRouter>
        );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('input').prop('value') ).toBe('batman');
    });
    
    test('debe mostrar un mensaje de error si no se encuentra el heroe', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search?q=batman1234'] }>
                <SearchScreen />
            </MemoryRouter>
        );

        expect( wrapper.find('.alert-danger').exists() ).toBe( true );
        expect( wrapper.find('.alert-danger').text().trim() ).toBe('No hay resultados para: batman1234');
    });
    
    test('debe llamar el navigate a la nueva pantalla (nuevo url)', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search'] }>
                <SearchScreen />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: 'batman'
            }
        });

        wrapper.find('form').prop('onSubmit')({
            // preventDefault(){}
            preventDefault: () => {}
        })

        expect( mockNavigate ).toHaveBeenCalledWith('?q=batman');
        
    });
    
    
});