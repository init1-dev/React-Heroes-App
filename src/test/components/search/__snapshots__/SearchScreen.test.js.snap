// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <SearchScreen /> debe mostrar a Batman y el input con el valor del queryString 1`] = `
Array [
  <h1
    className="mt-3"
  >
    Search
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar un Héroe"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-primary mt-1"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7 mb-5"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <a
        className="cards-link"
        href="/hero/dc-batman"
        onClick={[Function]}
      >
        <div
          className="col animate__animated animate__fadeIn mt-3"
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Batman"
                  className="card-img"
                  src="/assets/dc-batman.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-title"
                  >
                    Batman
                  </h5>
                  <p
                    className="card-text"
                  >
                    Bruce Wayne
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    >
                      Detective Comics #27
                    </small>
                  </p>
                  <p>
                    Click para ver más
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </a>
    </div>
  </div>,
]
`;

exports[`pruebas en <SearchScreen /> debe mostrarse correctamente con valores por defecto 1`] = `
Array [
  <h1
    className="mt-3"
  >
    Search
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar un Héroe"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-1"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7 mb-5"
    >
      <div
        className="alert alert-info"
      >
         Introduce un héroe a buscar 
      </div>
    </div>
  </div>,
]
`;
